planner/core/plan_cache_lru.go

```diff
@@ -0,0 +1,190 @@
+	bucket, bucketExist := l.buckets[string(key.Hash())]
+	if bucketExist {
+		if element, _, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			l.cache.MoveToFront(element)
+			return element.Value.(*CacheEntry).PlanValue, true
+		}
+	}
+	return nil, false
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *LRUPlanCache) Put(key kvcache.Key, value kvcache.Value, paramTypes interface{}) {
+	hash := string(key.Hash())
+	bucket, bucketExist := l.buckets[hash]
+	// bucket exist
+	if bucketExist {
+		if element, _, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			element.Value.(*CacheEntry).PlanValue = value
+			l.cache.MoveToFront(element)
+			return
+		}
+	}
+	// bucket not exist
+	newCacheEntry := &CacheEntry{
+		PlanKey:   key,
+		PlanValue: value,
+	}
+	element := l.cache.PushFront(newCacheEntry)
+	bucket = append(bucket, element)
+	l.buckets[hash] = bucket
+	l.size++
+	if l.size > l.capacity {
+		l.RemoveOldest()
+	}
+}
+
+// Delete deletes the multi-values from the LRU Cache.
+func (l *LRUPlanCache) Delete(key kvcache.Key) {
+	hash := string(key.Hash())
+	bucket, bucketExist := l.buckets[hash]
+	if bucketExist {
+		for _, element := range bucket {
+			l.cache.Remove(element)
+			l.size--
+		}
+		l.buckets[hash] = nil
+	}
+}
+
+// DeleteAll deletes all elements from the LRU Cache.
+func (l *LRUPlanCache) DeleteAll() {
+	for lru := l.cache.Back(); lru != nil; lru = l.cache.Back() {
+		l.cache.Remove(lru)
+		l.size--
+	}
+	l.buckets = make(map[string][]*list.Element)
+}
+
+// Size gets the current cache size.
+func (l *LRUPlanCache) Size() int {
+	return int(l.size)
+}
+
+// Values return all values in cache.
+func (l *LRUPlanCache) Values() []kvcache.Value {
+	values := make([]kvcache.Value, 0, l.cache.Len())
+	for ele := l.cache.Front(); ele != nil; ele = ele.Next() {
+		value := ele.Value.(*CacheEntry).PlanValue
+		values = append(values, value)
+	}
+	return values
+}
+
+// Keys return all keys in cache.
+func (l *LRUPlanCache) Keys() []kvcache.Key {
+	keys := make([]kvcache.Key, 0, len(l.buckets))
+	for _, bucket := range l.buckets {
+		keys = append(keys, bucket[0].Value.(*CacheEntry).PlanKey)
+	}
+	return keys
+}
+
+// SetCapacity sets capacity of the cache.
+func (l *LRUPlanCache) SetCapacity(capacity uint) error {
+	if capacity < 1 {
+		return errors.New("capacity of lru cache should be at least 1")
+	}
+	l.capacity = capacity
+	for l.size > l.capacity {
+		l.RemoveOldest()
+	}
+	return nil
+}
+
+// RemoveOldest removes the oldest element from the cache.
+func (l *LRUPlanCache) RemoveOldest() {
+	if l.size > 0 {
+		lru := l.cache.Back()
+		if l.onEvict != nil {
+			l.onEvict(lru.Value.(*CacheEntry).PlanKey, lru.Value.(*CacheEntry).PlanValue)
+		}
+
+		l.cache.Remove(lru)
+		l.removeFromBucket(lru)
+		l.size--
+	}
+}
+
+// removeFromBucket remove element from bucket
+func (l *LRUPlanCache) removeFromBucket(element *list.Element) {
```

Comment:
The complexity of this function is `O(n)` where n is the number of elements in this bucket. Can we optimize it to `O(1)`?

Reply:
yes, maybe map works, and we need a extra key